05-AdvancedTopics/web-search-mcp/README.md

```diff
@@ -0,0 +1,214 @@
+            # Call tools with your custom parameters
+            result = await session.call_tool("general_search", 
+                                           arguments={"query": "your custom query"})
+            # Process the result
+```
+
+## Tool Descriptions
+
+### general_search
+
+Performs a general web search and returns formatted results.
+
+**Parameters:**
+- `query` (string): The search query
+- `num_results` (integer, optional): Number of results to return (default: 5)
+
+**Example Request:**
+```json
+{
+  "query": "latest AI trends"
+}
+```
+
+### news_search
+
+Searches for recent news articles related to a query.
+
+**Parameters:**
+- `query` (string): The search query
+- `num_results` (integer, optional): Number of news articles to return (default: 5)
+
+**Example Request:**
+```json
+{
+  "query": "AI policy updates"
+}
+```
+
+### product_search
+
+Searches for products matching a query.
+
+**Parameters:**
+- `query` (string): The product search query
+- `num_results` (integer, optional): Number of product results to return (default: 5)
+
+**Example Request:**
+```json
+{
+  "query": "best AI gadgets 2025"
+}
+```
+
+### qna
+
+Gets direct answers to questions from search engines.
+
+**Parameters:**
+- `question` (string): The question to find an answer for
+
+**Example Request:**
+```json
+{
+  "question": "what is artificial intelligence"
+}
+```
+
+## Advanced Concepts Demonstrated
+
+1. **Multi-tool Orchestration**: Managing multiple tools with distinct purposes within a single server
+2. **API Rate Limit Handling**: Built-in timeout handling for external API calls
+3. **Structured Data Parsing**: Converting complex API responses into formatted, LLM-friendly text
+4. **Error Recovery**: Graceful degradation when API calls fail
+5. **Parameter Validation**: Proper handling of optional parameters with defaults
+
+## Troubleshooting
+
+### Common Issues
+
+1. **Missing SERPAPI_KEY in .env file**
+   - Error: `SERPAPI_KEY environment variable not found`
+   - Solution: Create or edit the `.env` file in the project root directory and add your SerpAPI key
+
+2. **Module not found errors**
+   - Error: `ModuleNotFoundError: No module named 'httpx'` (or similar)
+   - Solution: Ensure you've installed all dependencies using `pip install -r requirements.txt`
+
+3. **Connection issues**
+   - Error: `Error during client execution`
+   - Solution: Make sure the client and server are compatible versions and that server.py exists in the same directory
+
+4. **SerpAPI errors**
+   - Error: `Search API returned error status: 401`
+   - Solution: Verify your SerpAPI key is correct and not expired
+
+### Debug Mode
+
+To enable more verbose logging, set the logging level to DEBUG:
+
+```python
+# At the top of your client.py or server.py
+logging.basicConfig(
+    level=logging.DEBUG,  # Change from INFO to DEBUG
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
+)
+```
+
+## Why in 05-AdvancedTopics?
```

Comment:
I don't think we need this, it feels like it belongs as a comment in the PR.. But you're right, these are all good things to demonstrate :)